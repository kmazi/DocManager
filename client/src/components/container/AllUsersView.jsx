import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import AllUsersView from '../presentation/AllUsersView';
import { deactivateUser, fetchAllUsers } from '../../actions/userActions';

const mapStateToProps = state => ({
  allUsers: state.fetchAllUsers.users,
  responseStatus: state.fetchAllUsers.responseStatus,
  currentPage: state.fetchAllUsers.currentPage,
  counter: state.fetchAllUsers.counter,
  error: state.fetchAllUsers.status,
  deactivatedUserId: state.deactivateUser.deactivatedId,
});

export default connect(mapStateToProps,
  { deactivateUser, fetchAllUsers })(withRouter(AllUsersView));
