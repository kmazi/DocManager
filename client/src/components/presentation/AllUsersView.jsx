import React from 'react';
import PropTypes from 'prop-types';
import Alert from 'sweetalert2';
import Pagination from 'react-js-pagination';

const processRole = (roleId) => {
  switch (roleId) {
  case 1:
    return 'SuperAdmin';
  case 2:
    return 'Admin';
  case 3:
    return 'Fellow';
  case 4:
    return 'Learning';
  case 5:
    return 'Devops';
  default:
    break;
  }
};

const formatDate = (datetime) => {
  if (datetime) {
    const date = datetime.substring(0, datetime.indexOf('T'));
    const formattedDate = new Date(date);
    return formattedDate.toDateString();
  }
};

const AllUsersView = ({ allUsers, responseStatus, error,
  deactivateUser, counter, currentPage,
  fetchAllUsers }) => {
  const users = allUsers.map(user => (
    <tr id={user.id} key={user.id}>
      <td>
        <span>{user.username}</span>
      </td>
      <td>
        <span>{user.email}</span>
      </td>
      <td>
        <span>{processRole(user.roleId)}</span>
      </td>
      <td>
        <span>{formatDate(user.createdAt)}</span>
      </td>
      <td>
        <button
          id="useractionbtn"
          onClick={(event) => {
            event.preventDefault();
            Alert({
              title: `Comfirm ${user.isactive
        ? 'Deactivation' : 'Activation'}`,
              text: `Are you sure you want to
              ${user.isactive
        ? 'Deactivate' : 'Activate'} ${user.username}`,
              type: 'info',
              showCloseButton: true,
              showCancelButton: true,
              confirmButtonText:
                '<i class="fa fa-thumbs-up"></i> Yes!',
              cancelButtonText:
                '<i class="fa fa-thumbs-down"></i> No',
              showLoaderOnConfirm: true,
              preConfirm: () => new Promise((resolve, reject) => {
                deactivateUser(user.id)
                .then((res) => {
                  if (res.status === 'successful') {
                    resolve();
                  } else {
                    reject();
                  }
                });
              }),
              allowOutsideClick: false
            }).then(() => {
              Alert({
                type: 'success',
                title: `${user.isactive
                ? 'Activation' : 'Deactivation'} successful!`,
              });
            });
          }}
        >{user.isactive ? 'Deactivate' : 'Activate'}</button>
      </td>
      {/* <td>
        <button
          onClick={(event) => {
            event.preventDefault();
          }}
        >Change Role</button>
      </td> */}
    </tr>
  ));

  const userDetailsTable = (
    <div id="viewusers">
      <table className="table">
        <thead>
          <tr>
            <th>
              <span>UserName</span>
            </th>
            <th>
              <span>UserEmail</span>
            </th>
            <th>
              <span>Department/Role</span>
            </th>
            <th>
              <span>Registered on</span>
            </th>
          </tr>
        </thead>
        <tbody>
          {users}
        </tbody>
      </table>
      <div className={counter > 8 ? 'row' : 'row hide'}>
        <Pagination
          activePage={currentPage}
          itemsCountPerPage={8}
          totalItemsCount={counter}
          pageRangeDisplayed={5}
          onChange={(pageNumber) => {
            const offSet = (pageNumber - 1) * 8;
            fetchAllUsers(offSet, pageNumber);
          }}
        />
      </div>
    </div>
    );
  const errorMessage = <p>{error}</p>;
  return (<section>
    {responseStatus === 'successful' ? userDetailsTable : errorMessage}
  </section>
  );
};

AllUsersView.propTypes = {
  allUsers: PropTypes.arrayOf(PropTypes.shape({
    username: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    roleId: PropTypes.number.isRequired,
    createdAt: PropTypes.string.isRequired,
  })).isRequired,
  counter: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  error: PropTypes.string.isRequired,
  fetchAllUsers: PropTypes.func.isRequired,
  responseStatus: PropTypes.string.isRequired,
  deactivateUser: PropTypes.func.isRequired,
};

export default AllUsersView;
