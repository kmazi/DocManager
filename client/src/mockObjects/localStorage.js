/**
 * Mocks localstorage
 */
class LocalStorageMock {
  /**
   * the constructor
   */
  constructor() {
    this.store = {};
  }
/**
 * clears the localstorage
 * @return {void} returns null
 */
  clear() {
    this.store = {};
  }

/**
 * get item in the localstorage
 * @param {string} key - The key to the item in the localstorage
 * @return {string} returns the key to the item in localstorage
 */
  getItem(key) {
    return this.store[key];
  }
/**
 * set the value of an item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @param {string} value - The value of the item in the localstorage
 * @return {null} returns null
 */
  setItem(key, value) {
    this.store[key] = value;
  }

  /**
 * remove item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @return {void} returns nothing
 */
  removeItem(key) {
    delete this.store[key];
  }
}

global.localStorage = new LocalStorageMock();
export default global.localStorage;
