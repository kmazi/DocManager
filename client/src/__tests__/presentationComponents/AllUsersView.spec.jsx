import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import AllUsersView from '../../components/presentation/AllUsersView';
import '../../mockObjects/localStorage';

describe('The AllUsersView component:', () => {
  const response = { status: 'successful' };
  const props = {
    allUsers: [{ id: 1,
      username: 'jackson',
      email: 'jackson@gmail.com',
      roleId: 3,
      isactive: true,
      createdAt: Date(), }],
    responseStatus: 'successful',
    error: 'Access denied!',
    deactivateUser: jest.fn(() => Promise.resolve(response)),
    counter: 1,
    currentPage: 1,
    fetchAllUsers: jest.fn(() => Promise.resolve(response)),
  };
  test('should render correctly', () => {
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should render error message when responseStatus is unsuccessful',
  () => {
    props.responseStatus = 'unsuccessful';
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should render pagination link when counter is greater than 8',
  () => {
    props.counter = 9;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should display roleType as SuperAdmin when roleId is 1',
  () => {
    props.allUsers[0].roleId = 1;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should display roleType as Admin when roleId is 2',
  () => {
    props.allUsers[0].roleId = 2;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should display roleType as Learning when roleId is 4',
  () => {
    props.allUsers[0].roleId = 4;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should display roleType as Devops when roleId is 5',
  () => {
    props.allUsers[0].roleId = 5;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should not display roleType when roleId is not btw 1 and 5 inclusive',
  () => {
    props.allUsers[0].roleId = 9;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should button text to Activate when isactive prop is false',
  () => {
    props.allUsers[0].isactive = false;
    const enzymeWrapper = mount(<AllUsersView {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  // test('should ',
  //   () => {
  //     const enzymeWrapper = mount(<AllUsersView {...props} />);
  //     enzymeWrapper.find('Pagination')
  //     .props().onChange(2);
  //     const component = enzymeWrapper;
  //     const tree = toJson(component);
  //     expect(tree).toMatchSnapshot();
  //   });
});
