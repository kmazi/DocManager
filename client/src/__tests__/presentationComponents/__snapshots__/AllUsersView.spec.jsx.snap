// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AllUsersView component: should button text to Activate when isactive prop is false 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": false,
        "roleId": 9,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should display roleType as Admin when roleId is 2 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 2,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should display roleType as Devops when roleId is 5 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 5,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should display roleType as Learning when roleId is 4 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 4,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should display roleType as SuperAdmin when roleId is 1 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 1,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should not display roleType when roleId is not btw 1 and 5 inclusive 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 9,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should render correctly 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 3,
        "username": "jackson",
      },
    ]
  }
  counter={1}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="successful"
>
  <section>
    <div
      id="viewusers"
    >
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              <span>
                UserName
              </span>
            </th>
            <th>
              <span>
                UserEmail
              </span>
            </th>
            <th>
              <span>
                Department/Role
              </span>
            </th>
            <th>
              <span>
                Registered on
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id={1}
          >
            <td>
              <span>
                jackson
              </span>
            </td>
            <td>
              <span>
                jackson@gmail.com
              </span>
            </td>
            <td>
              <span>
                Fellow
              </span>
            </td>
            <td>
              <span>
                Invalid Date
              </span>
            </td>
            <td>
              <button
                id="useractionbtn"
                onClick={[Function]}
              >
                Deactivate
              </button>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        className="row hide"
      >
        <Pagination
          activePage={1}
          firstPageText="«"
          innerClass="pagination"
          itemsCountPerPage={8}
          lastPageText="»"
          nextPageText="⟩"
          onChange={[Function]}
          pageRangeDisplayed={5}
          prevPageText="⟨"
          totalItemsCount={1}
        >
          <ul
            className="pagination"
          >
            <Page
              activeClass="active"
              disabledClass="disabled"
              isActive={false}
              isDisabled={true}
              onClick={[Function]}
              pageNumber={1}
              pageText="«"
            >
              <li
                className="disabled"
                onClick={[Function]}
              >
                <a
                  className=""
                  href="#"
                >
                  «
                </a>
              </li>
            </Page>
            <Page
              activeClass="active"
              disabledClass="disabled"
              isActive={false}
              isDisabled={true}
              onClick={[Function]}
              pageNumber={0}
              pageText="⟨"
            >
              <li
                className="disabled"
                onClick={[Function]}
              >
                <a
                  className=""
                  href="#"
                >
                  ⟨
                </a>
              </li>
            </Page>
            <Page
              activeClass="active"
              disabledClass="disabled"
              isActive={false}
              isDisabled={true}
              onClick={[Function]}
              pageNumber={2}
              pageText="⟩"
            >
              <li
                className="disabled"
                onClick={[Function]}
              >
                <a
                  className=""
                  href="#"
                >
                  ⟩
                </a>
              </li>
            </Page>
            <Page
              activeClass="active"
              disabledClass="disabled"
              isActive={false}
              isDisabled={true}
              onClick={[Function]}
              pageNumber={1}
              pageText="»"
            >
              <li
                className="disabled"
                onClick={[Function]}
              >
                <a
                  className=""
                  href="#"
                >
                  »
                </a>
              </li>
            </Page>
          </ul>
        </Pagination>
      </div>
    </div>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should render error message when responseStatus is unsuccessful 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 3,
        "username": "jackson",
      },
    ]
  }
  counter={1}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;

exports[`The AllUsersView component: should render pagination link when counter is greater than 8 1`] = `
<AllUsersView
  allUsers={
    Array [
      Object {
        "createdAt": "Wed Aug 23 2017 10:03:42 GMT+0100 (WAT)",
        "email": "jackson@gmail.com",
        "id": 1,
        "isactive": true,
        "roleId": 3,
        "username": "jackson",
      },
    ]
  }
  counter={9}
  currentPage={1}
  deactivateUser={[Function]}
  error="Access denied!"
  fetchAllUsers={[Function]}
  responseStatus="unsuccessful"
>
  <section>
    <p>
      Access denied!
    </p>
  </section>
</AllUsersView>
`;
